Adding indenttabchar, indentspacechar and spacechar options

Neko-Box-Coder opened the pull request:
This PR added the options of `indentspacechar` and `spacechar` and moved `indentchar` to `indenttabchar` to avoid confusion.

`indenttabchar` will display the specified character when `\t` is encountered.

`indentspacechar` will display the specified character when `' '` is encountered at `tabsize` locations

`spacechar` will display the specified character when `' '` is encountered, but `indentspacechar` takes precedence.

Created this PR because I wanted the feature to view files with deeply nested blocks. 
Gonna merge it to my own fork https://github.com/Neko-Box-Coder/micro-dev for whoever needs it while waiting for this to be merged.

This should address long-standing requests of #2539 , #2707 and #431 

I am happy to revert the prompt message of telling user `indentchar` is deprecated, if people/maintainer finds it controversial.

JoeKar commented on internal/config/settings.go:
```diff
@@ -295,6 +310,7 @@ func InitGlobalSettings() error {
 // based on whether the path matches a glob
 // Must be called after ReadSettings
 func UpdatePathGlobLocals(settings map[string]interface{}, path string) {
+	checkDeprecatedSettings(settings)
```
Do we really need to check this here once again?
We've validated and maybe extended the settings (even the maps) in `validateParsedSettings()` already.

JoeKar commented on internal/config/settings.go:
```diff
@@ -311,6 +327,7 @@ func UpdatePathGlobLocals(settings map[string]interface{}, path string) {
 // based on whether the filetype matches to "ft:"
 // Must be called after ReadSettings
 func UpdateFileTypeLocals(settings map[string]interface{}, filetype string) {
+	checkDeprecatedSettings(settings)
```
Same as above.

dmaluka commented:
Do we need to deprecate `indentchar`? I'd rather fix it to do something closer to what its documentation says, instead of what it does now (i.e. simply displays all tab characters regardless of indentation).

It feels that by adding that many new options we are creating even more mess and confusion.

Why not just add `tabchar` which simply displays all tabs (i.e. what `indentchar` does now), `spacechar` which simply displays all spaces, and repurpose `indentchar`, for example, to display all tabs and all spaces at the position of tabs (i.e. divisible by tabsize)? And for example let `tabchar` (if set to any other value than ` `) take precedence over `indentchar`, so that if you want different characters for tabs and for spaces at the position of tabs, you can set `tabchar` for the former and `indentchar` for the latter?

BTW another request in https://github.com/zyedidia/micro/issues/431#issuecomment-382627101 was to display other non-printable characters, which is quite reasonable, so we might add another option e.g. `invisiblechar` for that.

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -203,12 +203,22 @@ Here are the available options:
-   setting available in many other text editors. The color of this character is
-   determined by the `indent-char` field in the current theme rather than the
-   default text color.
+* `indenttabchar`: sets the indentation character for tabs. This will not be
```
I guess technically they are the same? Since the tab character is always (displayed) aligned to `tabsize`?

I can change it to `tabsize`, I named it this ways just (trying) to be consistent to what it was called previously.

dmaluka commented on runtime/help/options.md:
```diff
@@ -203,12 +203,22 @@ Here are the available options:
+    default value: ` ` (space)
+
+* `indentspacechar`: same as `indenttabchar` except for spaces that are at
+   locations that are divisible by `tabsize`.
```
The problem is that the default behavior is confusing. For example, the default value of `tabsize` is 4, the user edits a file with 2-space indentation and doesn't bother changing the `tabsize` value (because it doesn't matter to him), then the user sets this `indentspacechar` option to highlight indentation, but only half of the indentation is highlighted.

I have no better suggestions right now, so I'm fine with it.

Neko-Box-Coder commented:
> > Repurposing a setting would be confusing since the user is not expecting it unless they read the documentation.
> 
> It would be "backward compatible" in a sense: tabs would be still highlighted.
> 
> Spaces at the position of tabs would become highlighted as well, but if the user changed the default value of `indentchar` in the first place, quite likely the user actually wants that?

Okay, I see. But if `tabchar` is "overriding" `indentchar` while `spacechar` isn't, wouldn't it make more sense to call it `indenttabchar`? And this hidden behavior (where `tabchar` is "overriding" `indentchar`) is no longer hidden?

I am fine to do it either way tbh. Just a question. 

The reason I deprecated `indentchar` other than clarity is because I don't need to check multiple settings for the same character (`tabchar` vs `indentchar` if both are populated) each time, makes it a bit easier to follow the logic in code as well. 

Another reason is what if the user actually wants space (the default) instead of `indentchar`, we wouldn't be able to tell as well.

> > I guess if we want to truly do what the name says, we can have a setting `indenttabchar` for all the tab characters before the first letter in a line and another setting `tabchar` for all the tabs after the first character, similar to how `indentspacechar` is handled.
> 
> Let's not overcomplicate without real need.

Okay.

dmaluka commented:
> But if `tabchar` is "overriding" `indentchar` while `spacechar` isn't, wouldn't it make more sense to call it `indenttabchar`?

Its primary role is to simply specify which character to use to visualize tabs. It is not primarily about indentation. What do we gain by making its name more verbose?



> And this hidden behavior (where `tabchar` is "overriding" `indentchar`)

It doesn't need to be hidden, it can (and probably should) be documented as well.

Neko-Box-Coder commented:
> > But if `tabchar` is "overriding" `indentchar` while `spacechar` isn't, wouldn't it make more sense to call it `indenttabchar`?
> 
> Its primary role is to simply specify which character to use to visualize tabs. It is not primarily about indentation. 

I know, but that's why it was called. 

That's why I said maybe we could have 2 settings for tabs, 1 act like `indentspacechar` and the other one act like `spacechar`.

> ... What do we gain by making its name more verbose?

To be less confusing. The previous behavior of `indentchar` is only for tabs. The user updates to latest micro and found out `indentchar` is now applied to indented space as well. If they want to change or disable it, they will have to unintuitively change the `tabchar` option (*remember, they want to change indented space now*) or mistakely think `spacechar` is the option they are looking for.

> > And this hidden behavior (where `tabchar` is "overriding" `indentchar`)
> 
> It doesn't need to be hidden, it can (and probably should) be documented as well.

I would prefer the behavior to be mentioned in the name rather that requiring the user to read the documentation (although they should be to be fair).


This is gonna be my last attempt to see if we can stick to the current proposed names.
If you still disagree, I can change it.

dmaluka commented:
Simple question: do we want an option that allows the user to simply display all tabs (no matter at the beginning, at the end or in the middle of a line)?

If the answer is yes, why not call this option `tabchar`? Wouldn't that be the most natural and intuitive name? If we name it `indenttabchar` instead, wouldn't _that_ be confusing?

If what we want instead (or in addition) is an option that only displays tabs at the beginning of lines, then yes, it would make sense to name it `indenttabchar`. But that is not what your PR is implementing.

Neko-Box-Coder commented:
> Simple question: do we want an option that allows the user to simply display all tabs (no matter at the beginning, at the end or in the middle of a line)?
> 
> If the answer is yes, why not call this option `tabchar`? Wouldn't that be the most natural and intuitive name? If we name it `indenttabchar` instead, wouldn't _that_ be confusing?

That's what it does currently, both `indentchar` and this PR's `indenttabchar`. The reason I keep the `indent` name is to be consistent with previous name, even though technically it acts like `spacechar`

> If what we want instead (or in addition) is an option that only displays tabs at the beginning of lines, then yes, it would make sense to name it `indenttabchar`. But that is not what your PR is implementing.

That's why I was asking maybe if I should have `indenttabchar` act like `indentspacechar` and `tabchar` that is used for in the middle of a sentence. But as you said it introduces complexity.

Actually, maybe it is better if I rename the new options to the following:
- `tabchar` which will replace (and act like) `indentchar`
- `indentspacechar` which highlights spaces that are in `tabsize` at the beginning of the sentence
- and `spacechar` for the rest of spaces
- and `invisiblechar` for non-displayable characters

Do you think this is better?

dmaluka commented:
> The reason I keep the `indent` name is to be consistent with previous name

What's the point of being consistent with it if you are deprecating it?



> That's why I was asking maybe if I should have `indenttabchar` act like `indentspacechar` and `tabchar` that is used for in the middle of a sentence. I can implement this instead But as you said it introduces complexity.

I'm not suggesting to implement it, I'm saying that _if_ we wanted to implement it, then `indenttabchar` would be a suitable name for it.

> Actually, maybe it is better if I rename the new options to the following:
> 
> [...]
> 
>    * `indentspacechar` which highlights spaces that are in `tabsize` at the beginning of the sentence

Ok, I've just noticed that you already have `bloc.X < leadingwsEnd` for `indentspacechar`...

Question (not a rhetorical one): are _you_ going to use `tabchar` and `indentspacechar` with different values, or with the same value?

Neko-Box-Coder commented:
> > Actually, maybe it is better if I rename the new options to the following:
> > [...]
> > 
> > * `indentspacechar` which highlights spaces that are in `tabsize` at the beginning of the sentence
> 
> Ok, I've just noticed that you already have `bloc.X < leadingwsEnd` for `indentspacechar`...
> 
> Question (not a rhetorical one): are _you_ going to use `tabchar` and `indentspacechar` with different values, or with the same value?

Yeah, I am, so that I can tell any hidden tabs in space-indented files.

dmaluka commented:
> so that I can tell any hidden tabs in space-indented files

For that purpose, wouldn't `hltaberrors` be more useful?

Neko-Box-Coder commented:
> > so that I can tell any hidden tabs in space-indented files
> 
> For that purpose, wouldn't `hltaberrors` be more useful?

True. I should probably turn that on as well.

The main thing here is we are mixing the ability to see tab character and indented blocks into a single option (`indentchar`), which now we are trying to separate.

If we reuse `indentchar`, this will certainly be confusing for people who are using `indentchar` just to see tab character, which now becomes `tabchar`. 

If we *are* reusing `indentchar`, I think we should at least have a new version/breaking changes message or something, similar to how lazygit did it when they introduced a breaking change. I really think just mentioning it in the documentation is not enough.

I think it is good to have a new version/breaking changes message anyway for future breaking changes if any.

dmaluka commented:
To be clear, when I suggested reusing `indentchar`, I thought your intention was to extend the definition of "indentation character" to all blank characters with `X % tabsize == 0`, not just those with `X < leadingWsEnd` (because I hadn't noticed your `bloc.X < leadingwsEnd` for `indentspacechar`, and because you confusingly didn't add it for `indenttabchar`).

> I think it is good to have a new version/breaking changes message anyway for future breaking changes if any.

This is a UI change, not an API change. The user is gonna notice it anyway.

We've done such "breaking" UI changes before. It is less breaking than https://github.com/zyedidia/micro/pull/3335, for example.



> Also if we need to account for a special case where `tabchar` can take precedence over `indentchar` when it is `' '`, why not `spacechar` as well?

Because with the "extended" definition of indentation character (see my confusion above), any tab is an indent char, but not any space is...

Anyway, screw that, yes, it would be messy. Now I think it's better to let `indentchar` take precedence over both `spacechar` and `tabchar`. So that, tabs at the end of a line are displayed as `tabchar`, but tabs (and spaces) at the beginning of a line are displayed as `indentchar` (unless it is set to `' '`). That would mean we can't set different characters for indentation tabs vs indentation spaces, but it seems there are no use cases for that?



> Or we can avoid all of this headache by just having 4 options, `spacechar`, `indentspacechar`, `tabchar` and `indenttabchar`

I'm almost fine with that, but... Think of what would be the most useful and pleasant behavior for the user. The user wants to see "indentation columns", displayed with the `|` character, for example. It seems friendlier to let the user simply set `indentchar` to `|` for that, instead of figuring out that it needs to be set separately for tabs and for spaces?

On the other hand... I've recalled once again that it's not that simple, since `tabwidth` may not match the actual indentation width with spaces. So yeah... for this reason it seems better to have separate options `indenttabchar` and `indentspacechar`, and in the documentation for `indentspacechar` emphasize that this option is tricky and the result will only be desirable if `tabwidth` is set appropriately...

So, right now I'm inclined to agree that having 4 options `spacechar`, `indentspacechar`, `tabchar` and `indenttabchar` would be best...





> and don't need to introduce breaking/confusing changes and special cases

It still doesn't answer what to do with `indentchar`.

I don't think silently removing it from the documentation (what your PR seems to be doing) is an option. It seems we should just keep it, and say in its documentation that it is just an alias for `tabchar`, kept for backward compatibility.

Neko-Box-Coder commented:
> > Similar to what vim does with `listchars`.
> 
> Hm, good idea. `micro` does something more "lightweight" with the `divchars` already, but this would be ugly to configure for more than two options.

I will need an example in order to implement grouping for the options. 

I am not following how `divchars` are grouped?.

I did say this PR addresses the issues but doesn't mean I am solving them completely (*for instance https://github.com/zyedidia/micro/issues/431#issuecomment-382627101 is asking for displaying a hex code, that's not gonna be done in this PR*). I am open to ideas but don't want to get feature-creeped here :sweat_smile: 

I will try to implement the ideas mentioned here if relatively straightforward, if not it's better to have a dedicated PR for those features.

The main goal for this PR (to me) is to have configurable indent characters for both space and tab indented files.

dmaluka commented:
> If we're going to keep adding more of these maybe grouping them together under one setting would be better for discoverability? Similar to what vim does with `listchars`.

Good point. Originally I was mildly opposed to doing something similar to vim's `listchars` (we don't need to exactly follow what vim does), but given that it would help us to avoid this "option creep", it seems attractive.

dmaluka commented:
> I will need an example in order to implement grouping for the options.

Something like `tab=>,space=.,itab=|,ispace=|` ?

I'm not sure what would be the name of this generic option though. `listchars` is definitely not the most intuitive one.

Neko-Box-Coder commented:
> I'm not sure what would be the name of this generic option though. `listchars` is definitely not the most intuitive one.

`formatchars` (welcome for any other suggestions)? Since space and tabs are used for formatting. But in this case newline (`eolchar`) would be in this option as well. `invchar` and `wrapchar` not so sure.


> Something like `tab=>,space=.,itab=|,ispace=|` ?

I see. In this case, I will need to store these options as global variables (since we don't want to do string parsing in every display call) instead of getting it from the settings map.

dmaluka commented:
> since we don't want to do string parsing in every display call

Hah, you are sure it would be slower than the settings map lookup? Especially considering _how many_ times we do the settings map lookup in every display call? :D

We are extremely wasteful there. See https://github.com/zyedidia/micro/issues/3227 and https://github.com/zyedidia/micro/issues/3228 (and https://github.com/zyedidia/micro/issues/2970 for the background)

(Just in case, I'm not saying we should not store those parsed settings.)

dmaluka commented:
`specialchars` maybe?

dmaluka commented:
> _for instance [#431 (comment)](https://github.com/zyedidia/micro/issues/431#issuecomment-382627101) is asking for displaying a hex code, that's not gonna be done in this PR_

Ah, I misread that comment, I thought it was just about displaying non-printable characters with some character... And I forgot that micro already does that, since a long time: https://github.com/zyedidia/micro/commit/51ab8f9914af6497f51cd3703328405418a61de5 + https://github.com/zyedidia/micro/commit/b793e9bb9234d63ded2b32386770503b88116976

JoeKar commented:
> `specialchars` maybe?

Maybe or `symbolchars`?

> for instance https://github.com/zyedidia/micro/issues/431#issuecomment-382627101 is asking for displaying a hex code, that's not gonna be done in this PR

Yes, that is ok, since we already show a replacement for non-displayable characters. Transforming them into their Unicode HEX would require more work unrelated to that.
So let us continue with the idea of grouping them into one string before we add four additional new options. The list within that one new option can be extended easily afterwards.
`tab=>,space=.,itab=|,ispace=|` is already a good option to start with. If the value for one isn't given then the default shall take place.

dmaluka commented:
> Or maybe `replacechars` (Inspired by `invisiblereplace`) since we are "replacing" the characters that match the condition with the ones we set?

That would be a pretty unconventional usage of the term "replace", especially in a text editor.

At the moment my favorite is `showchars`, it sounds very succinct and quite meaningful (meaning is: specify whether and how to show characters that are not shown otherwise).

JoeKar commented:
> At the moment my favorite is `showchars`

:+1:

Neko-Box-Coder commented:
I have just updated the PR to use `showchars` instead. There are a more few things I changed.

1. The show character value can be more than 1 char, in which case it will show the rest of them if there's space for it. 
    a. An example could be `itab=|>,tab=>,ispace=|` where an user can config `itab` to show both the pipe symbol `|` (to show scope) and the tab symbol `>`.
2. I have also removed `@` for wide rune. I was originally adding an option (`incomplete=@`) (I still have the change if we want to go that route) to set this char but I tried to play with different unicode characters and can't really see a scenario where rendering character other space is useful. This is related to #3516 so @JoeKar your opinion here would be great.

Neko-Box-Coder commented on internal/display/bufwindow.go:
```diff
@@ -511,37 +606,6 @@ func (w *BufWindow) displayBuffer() {
-								(!b.Settings["tabstospaces"].(bool) && bloc.X < leadingwsEnd && r == ' ' && !isTab) {
-								fg, _, _ := s.Decompose()
-								style = style.Background(fg)
-								dontOverrideBackground = true
```
Was in the middle of changing it, pushed it by accident. Should work now :)

dmaluka replied:
Still, what's the point of moving `hltaberrors` and `hltrailingws` out of `draw()`?

Neko-Box-Coder replied:
`hltaberrors` are related to space and tab characters themselves. It's weird to do the same character checks twice in different places.

I can kinda understanding moving `hltrailingws` back to `draw` since it doesn't check for the char.

It is a bit difficult to move things out from `draw` and deciding what should be moved and what shouldn't.
Right now (for the most part), draw "overrides" styles for non character specific things and the args are not as long as the initial suggestions.

dmaluka replied:
Well, maybe slimming down `draw()` is actually a step in the right direction, it is indeed way too big.

Let me take a closer look at your new version later, maybe it's actually better to keep it...

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -386,6 +382,22 @@ Here are the available options:
+   (This option overrides the `indentchar` option if that exists)
+   This option is specified in the form of `key1=value1,key2=value2,...`.
+   Here are the list of keys:
+   The color of the shown character is determined by the `indent-char`
```
I moved it to the bottom now.

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -203,14 +203,21 @@ Here are the available options:
 
     default value: `true`
 
-* `indentchar`: sets the indentation character. This will not be inserted into
```
Added it back with the shorter version.

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -203,14 +203,21 @@ Here are the available options:
+   in the file. This is similar to `listchars` in vim and the characters shown
+   will not be inserted into files. This option supersedes the `indentchar`
+   option.
+   The color of the shown character is determined by the `indent-char`
```
Updated

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -203,14 +203,21 @@ Here are the available options:
+   option.
+   The color of the shown character is determined by the `indent-char`
+   field in the current theme rather than the default text color.
+   This field is specified in the form of `option1=value1,option2=value2,...`.
```
Updated

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -203,14 +203,21 @@ Here are the available options:
+   field in the current theme rather than the default text color.
+   This field is specified in the form of `option1=value1,option2=value2,...`.
+   Here are the list of options:
+   - `ispace`: indent space characters (space before first visible character)
```
Yeah, typed it up but didn't read it again before pushing. Should read better now.

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -203,14 +203,21 @@ Here are the available options:
+   - `itab`: indent tab characters (tab before first visible character)
+   - `space`: space characters (space after first visible character)
+   - `tab`: tab characters (tab after first visible character)
+   An example of this option could be `tab=>,space=.,itab=|>,ispace=|`
```
Both read fine imo, updated.

Neko-Box-Coder commented on internal/display/bufwindow.go:
```diff
@@ -510,36 +624,8 @@ func (w *BufWindow) displayBuffer() {
-							}
-						}
+					if !bgoverridable {
+						dontOverrideBackground = true
```
True, missed that. Removed it.

dmaluka commented:
With the latest version:

- `ispace` is not working for me at all
- `tab` still not working for tabs at the beginning of line, even if `itab` is not set (looks like you've ignored my earlier comment about that?)

Neko-Box-Coder commented on internal/display/bufwindow.go:
```diff
@@ -494,133 +519,170 @@ func (w *BufWindow) displayBuffer() {
+				return r, style, bgoverridable
+			}
+
+			for _, mb := range matchingBraces {
```
Yeah, I thought the same as well but I tested it and seems good to me.

Neko-Box-Coder commented:
> With the latest version:
> 
> * `ispace` is not working for me at all

You sure?... It works for me though. (Before latest change)
This is what I have for testing, both works.
```json
//"showchars": "itab=▏⇥,ispace=▏,tab=⇥",
    "showchars": "ispace=▏",
```

---

> * `tab` still not working for tabs at the beginning of line, even if `itab` is not set (looks like you've ignored my earlier comment about that?)

Ah... My bad, must have missed it. You are right, probably more intuitive to have itab/ispace override space and tab. Just updated it now.

Neko-Box-Coder commented on internal/display/bufwindow.go:
```diff
@@ -494,133 +519,170 @@ func (w *BufWindow) displayBuffer() {
-							}
-						}
+		draw := func(r rune, combc []rune, style tcell.Style, highlight bool, showcursor bool, bgoverridable bool) {
+			defer func() {
```
I have just splitted it. The diffs are still horrible but I guess slightly better than before :sweat_smile:

dmaluka replied:
Why I actually commented on this: you've silently made yet another change here: reshuffling the code in `draw()` using `defer` (which you still have in the same commit as the rest of your changes in `draw()`). Although that is, IIUC, a simple non-functional change, yet its diff is not small (since it changes indentation in the entire `draw()`), and it messes with other changes you've made in `draw()`, making it quite impossible to reviewer them (unless the reviewer looks at github's diff for the force-push in which you added this change; assuming that the reviewer is willing to do such archaeology of your force-pushes in this PR). So it's kind of obvious it's better to do such change in a separate commit?

dmaluka replied:
Yeah, looks good, thanks.

dmaluka commented:
> You sure?...

Yes.

```
% ./micro -version
Version: 2.0.15-dev.219
Commit hash: 3068b4af
Compiled on July 13, 2025
```

and yet, `space` works, `ispace` doesn't. Even with an empty config dir.

dmaluka commented on runtime/help/options.md:
```diff
@@ -386,6 +382,24 @@ Here are the available options:
+               the first visible character
+   - `itab`: indent tab characters, for tab before the first visible character
+   - `space`: space characters, for space that is not displayed by `ispace`
+   - `tab`: tab characters, for tab that is not displayed by `itab`
```
1. Description outdated (`space` and `tab` are now for all spaces/tabs, not just "non-indent" ones)?
2. It's better to describe `space` and `tab` (i.e. the more "basic" case) before `ispace` and `itab`?

Neko-Box-Coder replied:
Updated

Neko-Box-Coder commented:
> With `snapcraft.yaml` I indeed see the same... Because micro's built-in `ftoptions` plugin enables `tabstospaces` for YAML files.
> 
> So, with `tabstospaces` set to true, `ispace` indeed works for me. With `tabstospaces` set to false (which is its default value), it doesn't.

`ispace` is **not** supposed to shown in a tab file (`tabstospaces`), `itab` is instead.

Okay, if you open up `actions.go` (`micro -config-dir ./EmptyDir ./internal/action/actions.go`) with the config I showed earlier, it should not show anything.

But if you set your config to this
```json
{
    "showchars": "ispace=|,space=.,itab=|,tab=>"
}
```
It should show `|` at leading tabs and `>` for non leading ones. Do you see the same? Or are we misunderstanding what the inner options of `showchars` supposed to do?

dmaluka commented:
Yes, I see the same, I was talking only about spaces, not tabs.

Ok, after some thinking, making it dependent on `tabstospaces` makes sense, it is also in line with the behavior of `hltaberrors` (which I myself invented, after all). But, it's better to document it as well?

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -386,6 +382,25 @@ Here are the available options:
+   - `space`: space characters
+   - `tab`: tab characters
+   - `ispace`: indent space characters, for space at indent position before
+               the first visible character if set
```
The `if set` was trying to convey `ispace` overrides `space`, I guess I wasn't explicit enough.

Just reworded it, does it read better now?

dmaluka commented:
Although, after yet more thinking: what's the _actual_ advantage of making it dependent on `tabstospaces`?

`hltaberrors` uses `tabstospaces` as a heuristic for deciding whether a tab or space character is an error or not. But why do `ispace` and `itab` need to care about that?

Neko-Box-Coder commented:
> Although, after yet more thinking: what's the _actual_ advantage of making it dependent on `tabstospaces`?
> 
> `hltaberrors` uses `tabstospaces` as a heuristic for deciding whether a tab or space character is an error or not. But why do `ispace` and `itab` need to care about that?

It's just more uniform to other existing micro features, as well as less convoluted logic in the code if we know if the file is space or tab indented.  I don't see much of a problem relying on `tabstospaces`, that being said the name of that option could be not so intuitive. But still, my point still stands I think.

Neko-Box-Coder commented:
> > Although, after yet more thinking: what's the _actual_ advantage of making it dependent on `tabstospaces`?
> > `hltaberrors` uses `tabstospaces` as a heuristic for deciding whether a tab or space character is an error or not. But why do `ispace` and `itab` need to care about that?
> 
> It's just more uniform to other existing micro features, as well as less convoluted logic in the code if we know if the file is space or tab indented. I don't see much of a problem relying on `tabstospaces`, that being said the name of that option could be not so intuitive. But still, my point still stands I think.

Also, I cba to change the logic again. The number of back and forth on this is already more than what I get normally at work :sweat_smile:  (which is a good thing, but just really cba xd)

dmaluka commented on runtime/help/options.md:
```diff
@@ -386,6 +382,27 @@ Here are the available options:
+   - `space`: space characters
+   - `tab`: tab characters
+   - `ispace`: space characters at indent position before the first visible
+               character in a line which overrides `space` if this is set.
```
> If this is not set, `space` will be shown instead.

Sounds good.

Neko-Box-Coder replied:
Cool, updated

Neko-Box-Coder commented:
> Which features besides hltaberrors (which, as I said, uses it as a heuristic ...

`retab` on top of my head.

> How? If we remove this dependency, it means we just remove the b.Settings["tabstospaces"].(bool) check, we don't add any complexity to the code?

True but it creates weird cases like what if you have a mixture of leading tabs and spaces?..

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -386,6 +382,27 @@ Here are the available options:
+   - `ispace`: space characters at indent position before the first visible
+               character in a line which overrides `space` if this is set.
+               This is only shown when `tabstospaces` is true.
+   - `itab`: for tab characters before the first visible character in a line
```
Updated.

dmaluka commented:
> what if you have a mixture of leading tabs and spaces?..

What's wrong with that? (I mean, not with the fact you have this mixture in the first place, but with the fact that we honestly display it as it is.)

Neko-Box-Coder commented:
> What's wrong with that? (I mean, not with the fact you have this mixture in the first place, but with the fact that we honestly display it as it is.)

Yes but a fair assumption is the user would expect indent char is displayed depending if the editor thinks you are in a space or tab indented file

So let's say I have `"showchars": "ispace=|,itab=|` for example to mimic scope indicator in other text editor

If the file has a mixture of leading spaces and tabs like this (assuming `tabsize` is 4) (either intentional or by accident, who knows)
```
<tab><space><space><space><space><tab><tab>abc
```

If we ignore `tabstospace`, it will look like
```
|   |   |   |   abc
```

If we don't ignore it, it will look like

```
|       |   |   abc
```

I would prefer the latter even if the differences are minimal.

I think showing an "indent" character when the editor thinks otherwise (`tabstospace`) is misleading.

Andriamanitra commented:
> So let's say I have `"showchars": "ispace=|,itab=|` for example to mimic scope indicator in other text editor
> 
> If the file has a mixture of leading spaces and tabs like this (assuming `tabsize` is 4) (either intentional or by accident, who knows)
> 
> ```
> <tab><space><space><space><space><tab><tab>abc
> ```
> 
> If we ignore `tabstospace`, it will look like
> 
> ```
> |   |   |   |   abc
> ```
> 
> If we don't ignore it, it will look like
> 
> ```
> |       |   |   abc
> ```
> 
> I would prefer the latter even if the differences are minimal.
> 
> I think showing an "indent" character when the editor thinks otherwise (`tabstospace`) is misleading.

Shouldn't it be like this, regardless of `tabstospaces`?
```
|   |||||   |   abc
```

Neko-Box-Coder commented:
> 
> Shouldn't it be like this, regardless of `tabstospaces`?
> 
> ```
> |   |||||   |   abc
> ```

No because `ispace` only gets displayed at columns divisible by `tabsize`. 

What you shown would be true if `space` is set to `|` as well.

dmaluka commented:
> So let's say I have `"showchars": "ispace=|,itab=|` for example to mimic scope indicator in other text editor

Hmm... this is why IIRC I was originally asking you if there are really use cases for having two separate "scope indicators" for tabs and spaces, rather than just one for both. You said you had use cases for that.

So... instead of requiring to the user to "mimic scope indicator in other text editor" by setting both `ispace` and `itab`, maybe we should, well, just provide this scope indicator? Name it e.g. `indent`, and make it respect `tabstospaces`, and then we can make `ispace` and `itab` not care about `tabstospaces`. So the user can just set `indent=|`, and then your example will be displayed as `|       |   |   abc`, as we want it to. While if the user sets e.g. `itab=|,ispace=:`, it will be always displayed as `|   :    |   |   abc`, as the user probably wants it to, instead of unexpectedly hiding `:` or `|` depending on the `tabstospaces` value.

Neko-Box-Coder commented:
@Andriamanitra 

> I see, that's pretty weird (does any other editor work like that?). I had incorrectly assumed `ispace` would be used for showing spaces in indentation but not in the middle of the line, kind of like a less smart version of `renderWhitespace: "boundary"` in VS Code.

For me, I just want to be able to see `tab` character without losing the scope indicator, that's why I needed 2 indent characters to do that.

VS Code is different because it can render a tab character on top of the scope line, we can't do that here since we can only render 1 character.

dmaluka commented:
> But anyway, the behavior that `indent` respects `tabstospaces` but not `ispace` (indent space) and `itab` (indent tab) is weird to me.

What's weird about it? Space indentation in a tab-indented file it still an indentation, just not a "desired" one.

> the fact that the user told the editor if the file is tab indented or not

The user hardly told anything. The `tabstospaces` option just happens to be set to one value or the other, not necessarily because the user has set it, and even if so, not necessarily because the user realized this particular implication of it (after all, this is not the primary, documented purpose of `tabstospaces`).

Neko-Box-Coder commented:
Yeah I dropped the `tabstospaces` check.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -514,28 +539,45 @@ func (w *BufWindow) displayBuffer() {
+					if indenttabchars != "" {
+						indentrunes = []rune(indenttabchars)
+					} else {
+						indentrunes = []rune(tabchars)
```
Instead of having this `else` branch which duplicates the outer `else` branch below, we can just add `indenttabchars != ""` to the outer `if` condition?

Ditto for `spacechars` below.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -514,28 +539,45 @@ func (w *BufWindow) displayBuffer() {
-					fg, _, _ := s.Decompose()
-					style = style.Foreground(fg)
+			case ' ':
+				if bloc.X%tabsize == 0 && bloc.X < leadingwsEnd {
```
Simply checking `bloc.X%tabsize == 0` is buggy, since `bloc.X` is not necessarily the _visual_ position of the character in the line (since there may be tab characters in the line before this space character). Try it: put some space characters after some tab characters (before the first visible character in a line) -> result: `ispace` is displayed at wrong locations.

We should probably use `util.StringWidth()`.

Neko-Box-Coder replied:
I am using the line width (`totalwidth`) now, doing `util.StringWidth()` every time would be overkill.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -514,28 +539,45 @@ func (w *BufWindow) displayBuffer() {
-					fg, _, _ := s.Decompose()
-					style = style.Foreground(fg)
-				}
+			var returnrune rune
```
nit: maybe `drawrune` instead (in line with the name used by the callers, and because the meaning of this variable is precisely: the rune to be drawn)?

Neko-Box-Coder replied:
Updated.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -494,7 +494,87 @@ func (w *BufWindow) displayBuffer() {
 
-		draw := func(r rune, combc []rune, style tcell.Style, highlight bool, showcursor bool) {
+		getRuneStyle := func(r rune, style tcell.Style, isplaceholder bool) (rune, tcell.Style, bool) {
+			bgoverridable := true
```
nit: `bgoverridable` remains true in almost all cases, except `hltaberrors` and `hltrailingws`. So maybe move its declaration closer to the point where is becomes useful, i.e. before the `hltaberrors` checks, and in all the earlier `return` statements just use `true`?

Neko-Box-Coder replied:
Moved and renamed it to preservebg. To me `override` makes sense since we are overriding the style, but either is fine I guess.

dmaluka replied:
> renamed it to preservebg.

But why in a [separate commit](https://github.com/zyedidia/micro/pull/3760/commits/90d911efe77ca7a199bb4d304a28882d61d29883)?

BTW that commit is broken (only the next commit fixes it):

```
internal/display/bufwindow.go:521:8: returnrune declared but not used
internal/display/bufwindow.go:575:11: undefined: drawrune
make: *** [Makefile:27: build-quick] Error 1
```

Neko-Box-Coder replied:
Merged the commits.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -509,37 +589,8 @@ func (w *BufWindow) displayBuffer() {
 
 					// syntax or hlsearch highlighting with non-default background takes precedence
 					// over cursor-line and color-column
```
nit: maybe move this comment inside the `if` (since it relates specifically to the `origBg == defBg` check)?

Neko-Box-Coder replied:
Updated

Neko-Box-Coder commented on runtime/help/options.md:
```diff
@@ -386,6 +382,26 @@ Here are the available options:
 
+* `showchars`: sets what characters to be shown to display various invisible
+   characters in the file. The characters shown will not be inserted into files.
+   (This option overrides the `indentchar` option if that exists)
```
I reworded it.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -494,28 +494,28 @@ func (w *BufWindow) displayBuffer() {
 		}
 		bloc.X = bslice
 
+		// returns the rune to be drawn, style of it and if the bg should be preserved
```
What's the point of this comment?

dmaluka replied:
That's what we do - look at the call sites to see how the function is actually used and what for. If we are not interested in the call sites, why would we be interested in the function itself?

If the problem is just that we don't see the _names_ of the return values here in the function declaration, we could use named return values? Or the problem is that we don't want to use them since the line would become enormously long (which it already is)?

Anyway, I'm ok with keeping it as is.

dmaluka replied:
But again, why are you doing that in this commit "Simplifying draw to be less nested" which is supposed to be just about simplifying `draw()` to be less nested? This comment should added in the first commit, where `getRuneStyle()` itself is added?

And same about renaming `bgoverridable` and `returnrune` - why do that in this commit, and why do that at all? If we agree they should be `preservebg` and `drawrune`, we should name them `preservebg` and `drawrune` from the beginning, in the first commit?

I genuinely don't get it.

Neko-Box-Coder replied:
It's fine, just merged to the wrong commit, should be good now?

dmaluka replied:
Yeah, now looks good.

Except that, the commit message of the 2nd commit includes an outdated sentence?

Neko-Box-Coder replied:
oops, missed that. Have removed it.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -597,10 +598,10 @@ func (w *BufWindow) displayBuffer() {
+				if !preservebg {
+					// syntax or hlsearch highlighting with non-default background takes precedence
+					// over cursor-line and color-column
+					preservebg = origBg != defBg
```
Also just realized: this can be even simpler:

```go
// syntax or hlsearch highlighting with non-default background takes precedence
// over cursor-line and color-column
if !preservebg && origBg != defBg {
	preservebg = true
}
```

Neko-Box-Coder replied:
Updated, both reads fine to me tbh.

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -495,7 +520,7 @@ func (w *BufWindow) displayBuffer() {
 
 		// returns the rune to be drawn, style of it and if the bg should be preserved
-		getRuneStyle := func(r rune, style tcell.Style, isplaceholder bool) (rune, tcell.Style, bool) {
+		getRuneStyle := func(r rune, style tcell.Style, drawoffset int, linex int, isplaceholder bool) (rune, tcell.Style, bool) {
```
Any reason why you renamed `showoffset` to `drawoffset`?
Neither name seems quite intuitive to me, but at least `showoffset` somehow relates to `showchars` (and `showchars` is the only reason for the existence of this argument, right?).

dmaluka commented on internal/display/bufwindow.go:
```diff
@@ -697,6 +731,7 @@ func (w *BufWindow) displayBuffer() {
 
 			width := 0
 
+			linex := totalwidth
```
`visualx`? (like in `VLocFromLoc()` for example)

dmaluka commented on runtime/help/options.md:
```diff
@@ -386,6 +382,25 @@ Here are the available options:
+
+   Here are the list of keys:
+   - `space`: space characters
+   - `tab`: tab characters, this option overrides the `indentchar` option
```
"this key"?
Or, cleaner, in a separate sentence: "If set, overrides the `indentchar` option."?

Neko-Box-Coder commented:
Hey @JoeKar 

I want to get this merged so would be nice if you can take a look at this PR and approve/merge it when you are free.

Thank.

Neko-Box-Coder commented on internal/display/bufwindow.go:
```diff
@@ -494,7 +494,88 @@ func (w *BufWindow) displayBuffer() {
+				}
+			}
+
+			if r != '\t' && r != ' '{
```
Fixed in the first commit now.

Neko-Box-Coder commented:
@JoeKar 
Maybe we should close #2539 and #2707 ?
